### ISSUE TITLE
Hyperparameter Transfer beyond MuP

### ISSUE THREAD

**Issue by plugyawn** [NVIDIA/Megatron-LM#4088]
**Is your feature request related to a problem? Please describe.**
#3058, #3715 introduces MuP into Megatron-LM with support for Muon.
MuP allows efficient and reliable hyperparameter (esp. LR) transfer from narrow to wide networks, for the same depth. 

There are a series of papers in this field, that allow transfer from shallow to deeper networks, etc, which are essential for a good pretraining scaling recipe. 

A new paper from Microsoft, HyperP, claims to be "the first framework for transferring optimal learning rates across model width, depth, training tokens, and Mixture-of-Experts (MoE) granularity under the Frobenius-sphere constraint with the Muon optimizer." 

I suggest we integrate HyperP into Megatron-LM.
Link: [Rethinking Language Model Scaling under Transferable Hypersphere Optimization](https://arxiv.org/pdf/2603.28743), Ren et al, 2026.

Tag the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) 
to get oncall's attention to this issue.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by plugyawn** [comment]
@BoxiangW @janEbert if it's fine, I have been testing out an implementation, could put in a PR, or we could decouple this into different axes: tokens, MoEs, etc, since the validation might be hard on a single PR for everything.

**Comment by plugyawn** [comment]
Their github repo seems to be in conflict with some of the claims on the paper (I was looking through the scaling on some of the matrices, they seem to not be discussed in the paper). We also need a MuonH/AdamH implementation, so perhaps this can be seen as a continuing series of additions.

**Comment by janEbert** [comment]
Hey, I would say if you already have an implementation and it's not too much work to clean up, it does make sense to PR. Ideally, we'd discuss the design a bit in advance, but I think we could then just do it on the practical example that you already have.

Purely personally, I'm also very interested in the [Scion optimizer](https://arxiv.org/abs/2502.07529) and its theory, for which its group has also developed a rule for scaling (batch size, seq len, LR) combinations across token dimensions:
https://arxiv.org/abs/2603.21191
And a similar paper on finding the optimal (batch size, LR) combination for Scion:
https://arxiv.org/abs/2510.03871

As for the modified optimizer implementations, [the Emerging Optimizers repository already seems to contain them](https://github.com/NVIDIA-NeMo/Emerging-Optimizers/blob/6f81030909f4bd82d922d91be777606b5a2ea5a2/emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py).

**Comment by plugyawn** [comment]
I think it'd be great to discuss the design before a PR too! (tried looking for a Discord/Zulip but couldn't find one, would love to join if one exists!) 
Also I'm pretty sure I need to implement it afresh to understand some of the parts better, and I haven't run longer/larger tests beyond 1024 dims/enough points to make sure things are working fine. 

Scion is great! I think perhaps we could turn this into a set of issues to tackle, to shape up the hyperparam transfer landscape for Megatron, and provide some pre-built scaling recipes cheaply? 

I just saw the Emerging-Optimizers implementation! I think it should be straightforward to wrap them in Megatron, same as Muon.

**Comment by janEbert** [comment]
Got it; in that case, let's discuss first. I think staying on GitHub is good, but if I hear something else, I'll let you know. :)
There are also some other people in other time zones who I'm sure are interested in this discussion as well.

> I think perhaps we could turn this into a set of issues to tackle, to shape up the hyperparam transfer landscape for Megatron, and provide some pre-built scaling recipes cheaply?

Yes, that sounds lovely! The pre-built scaling recipes have also already been a dream of mine! :D

And after all, this field is still fast-moving and not figured out. Nobody really knows what the best method to implement would be. My main concern for the codebase here is feature creep with trying to support a bunch of different optimization schemes. Since it's not really clear what the tradeoffs of the different schemes are, especially at larger scales, I'm worried of having to maintain a bunch of implementations that may become dusty soon after the next paper. If we can implement things in a general fashion, so that future algorithms could re-use these foundational parameters/functions and only require minor changes, I guess that would be the best for the codebase from a future perspective.

**Comment by mkhona-nvidia** [comment]
@plugyawn We are also looking at other versions of hyperparameter transfer such as recently published work on Spectral muP + depth scaling: https://arxiv.org/abs/2603.00541

**Comment by plugyawn** [comment]
> Yes, that sounds lovely! The pre-built scaling recipes have also already been a dream of mine! :D

Would love to help out! This would be very cool! A set of strong primitives that could ideally extend to Mamba, GDN/GDN2, etc with solid experimentation, I assume, would be very relevant. 

Re: feature creep; I think a short research programme is possible as of today that could possibly "unify" the different aspects of looking at MuP: I think Greg Yang's initial TPV MuP got superceded with the new derivation in TPVI, and as @mkhona-nvidia shared, https://arxiv.org/abs/2603.00541 looks interesting (I think some of this overlaps with #3177 too!)

There seems to be a slew of these results: I remember Complete(P) from Apple, and Hypersphere Optimization are recent ones that caught my eye. [kexue.fm's blog on MuP/Muon correspondence](https://kexue.fm/archives/11340) was also very nice, and collates a lot of the research as of October 2025. 

The latest "frontier jargon" seems to be shifted to finding similar rules for RL, so might be interesting to make Megatron a testbed for these scaling exercises, and develop a similar plan as RL scaling "tricks" also start appearing soon (Ian Osband has some interesting work out recently on what he calls the "delightful policy gradient". It appears have the bearings of an important paper in the field, and my initial experiments wrapping it over GRPO on Nemo/RL worked very well. While not strictly similarly to the HP transfer of MuP, a general "theory" of hyperparameter transfer over general learning algorithms over large models might soon be possible).

**Comment by plugyawn** [comment]
@janEbert @mkhona-nvidia do you have a set of ideas that you would like prototyped before moving ahead?

**Comment by janEbert** [comment]
I've found that the framework of [scaling exponents](https://arxiv.org/abs/2407.05872) still holds up well. Maybe a more fundamental implementation, basically a generalization of the μP implementation, that allows for varying these knobs via configuration flags makes sense as a first step in going beyond μP. This would lay a foundation and allow for much easier implementations of follow-up methods like what you and @mkhona-nvidia linked.

I'm not sure this is quite what you asked for, though. 😅

**Comment by plugyawn** [comment]
That's helpful! That sounds like the right way forward. Thanks for the paper! Seems to verify some things I was suspecting.

I'll go through it, and propose a PR trying to unify the current MuP scaling with allowance for depth, expert, etc scaling as well. 
Only caveat is that I usually spend on GPU compute myself, so I'm a bit limited there (but this sounds like a great learning opportunity, I don't mind as long as the team is okay with a few delays here and there!)

**Comment by plugyawn** [comment]
I suppose the first pass should be a general scaling-policy with the current #3058 MuP refactored as a preset, while exposing the key scaling knobs through a config layer? Depth-MuP, and (later later) HyperP (which will need some changes on Emerging-Optimizers for Muon/MuonH too) can be exposed via a similar preset to MuP.

**Comment by Phlip79** [comment]
Thank you @plugyawn, let us know once you have a PR ready please.

**Comment by plugyawn** [comment]
On it!

**Comment by plugyawn** [comment]
Will have the draft up by today.

**Comment by plugyawn** [comment]
@Phlip79 opened a PR at #4381! Waiting on the next set of reviews.

**Comment by janEbert** [comment]
Hey, no worries, this is on our radar. :)
